src/include/migraphx/operators.hpp

```diff
@@ -807,41 +807,204 @@ struct gather
+    {
+        if(b.empty())
+            return a;
+
+        if(a.empty())
+        {
+            return b;
+        }
+
+        auto a_size = a.size();
+        auto b_size = b.size();
+
+        auto n_dim = std::min(a_size, b_size);
+        std::vector<std::size_t> out_lens(std::max(a_size, b_size));
+        for(std::size_t i = 0; i < n_dim; ++i)
+        {
+            if(a[a_size - 1 - i] == b[b_size - 1 - i])
+            {
+                out_lens[i] = a[a_size - 1 - i];
+            }
+            else if(b[b_size - 1 - i] == 1)
+            {
+                out_lens[i] = a[a_size - 1 - i];
+            }
+            else if(a[a_size - 1 - i] == 1 && is_mutli_broadcast)
+            {
+                out_lens[i] = b[b_size - 1 - i];
+            }
+            else
+            {
+                MIGRAPHX_THROW("DOT : dimension mismatch, matrix A: {" + to_string_range(a) +
+                               "}, and matrix B: {" + to_string_range(b) +
+                               "} are not broadcastable");
+            }
+        }
+
+        if(a_size > n_dim)
+        {
+            std::copy(a.rbegin() + n_dim, a.rend(), out_lens.begin() + n_dim);
+        }
+
+        if(b_size > n_dim)
+        {
+            std::copy(b.rbegin() + n_dim, b.rend(), out_lens.begin() + n_dim);
+        }
+
+        std::reverse(out_lens.begin(), out_lens.end());
+
+        return out_lens;
+    }
+
     std::string name() const { return "dot"; }
     shape compute_shape(std::vector<shape> inputs) const
     {
+        // If there are 3 inputs, there are two scenarios:
+        // 1. A and B are 2-D matrices and C is broadcastable to A * B
+        // 2. A and B are stack of matrices, then shape for the batch
+        // should be the same for A and B, and C is the same shape
+        // as A * B (For now, we add this requirement to simplify the
+        // implementation. we can remove this requirement later)
+        if(inputs.size() == 3)
+        {
+            auto a_lens   = inputs[0].lens();
+            auto b_lens   = inputs[1].lens();
+            auto out_lens = a_lens;
+            auto t        = inputs[0].type();
+            if(a_lens.size() != b_lens.size() ||
+               !std::equal(a_lens.rbegin() + 2, a_lens.rend(), b_lens.rbegin() + 2))
+            {
+                MIGRAPHX_THROW("DOT: dimension mismatch, operand A: {" + to_string_range(a_lens) +
+                               "}, cannot multiply operand B: {" + to_string_range(b_lens) + "}");
+            }
+
+            std::size_t dim_0 = a_lens.size() - 2;
+            std::size_t dim_1 = a_lens.size() - 1;
+            if(a_lens[dim_1] != b_lens[dim_0])
+            {
+                MIGRAPHX_THROW("DOT: inner dimensions do not match, operand A: {" +
+                               to_string_range(a_lens) + "}, operand B: {" +
+                               to_string_range(b_lens) + "}");
+            }
+            out_lens[dim_1] = b_lens[dim_1];
```

Comment:
It might be cleaner to create a helper function check_dims(a_lens, b_lens) to check for the dimension mismatches

Reply:
fixed by adding a function to check the dims